parsec/core/gui/invite_user_widget.py

```diff
@@ -1,206 +1,276 @@
-    invite_and_create_user as core_invite_and_create_user,
-)
-from parsec.core.types import BackendOrganizationClaimUserAddr
-
-from parsec.core.backend_connection import BackendNotAvailable, BackendConnectionError
+from parsec.api.protocol import InvitationType, InvitationStatus, InvitationDeletedReason
+from parsec.core.types import BackendInvitationAddr
+
+from parsec.core.backend_connection import backend_authenticated_cmds_factory
 from parsec.core.gui import desktop
-from parsec.core.gui import validators
-from parsec.core.gui.custom_dialogs import show_info, show_error, GreyedDialog
+from parsec.core.gui.greet_user_widget import GreetUserWidget
+from parsec.core.gui.custom_dialogs import show_error, GreyedDialog, ask_question
 from parsec.core.gui.lang import translate as _
+from parsec.core.gui.trio_thread import JobResultError, ThreadSafeQtSignal, QtToTrioJob
 from parsec.core.gui.ui.invite_user_widget import Ui_InviteUserWidget
-from parsec.core.gui.trio_thread import JobResultError, ThreadSafeQtSignal
+from parsec.core.gui.ui.user_invitation_widget import Ui_UserInvitationWidget
 
 
 logger = get_logger()
 
 
-async def _do_registration(core, device, new_user_id, token, is_admin):
-    try:
-        new_user_id = UserID(new_user_id)
-    except ValueError as exc:
-        raise JobResultError("registration-invite-bad-value") from exc
+async def _do_invite_user(device, config, email):
+    async with backend_authenticated_cmds_factory(
+        addr=device.organization_addr,
+        device_id=device.device_id,
+        signing_key=device.signing_key,
+        keepalive=config.backend_connection_keepalive,
+    ) as cmds:
+        rep = await cmds.invite_new(type=InvitationType.USER, claimer_email=email, send_email=False)
+        if rep["status"] != "ok":
+            raise JobResultError(rep["status"])
+        action_addr = BackendInvitationAddr.build(
+            backend_addr=device.organization_addr,
+            organization_id=device.organization_id,
+            invitation_type=InvitationType.USER,
+            token=rep["token"],
+        )
+        return action_addr
+
 
-    try:
-        rep = await core.user_fs.backend_cmds.user_find(new_user_id)
-    except BackendNotAvailable as exc:
-        raise JobResultError("registration-invite-offline") from exc
-    except BackendConnectionError as exc:
-        raise JobResultError("registration-invite-error", info=str(exc)) from exc
+async def _do_list_invitations(device, config):
+    async with backend_authenticated_cmds_factory(
+        addr=device.organization_addr,
+        device_id=device.device_id,
+        signing_key=device.signing_key,
+        keepalive=config.backend_connection_keepalive,
+    ) as cmds:
+        rep = await cmds.invite_list()
+        if rep["status"] != "ok":
+            raise JobResultError(rep["status"])
+        return [inv for inv in rep["invitations"] if inv["type"] == InvitationType.USER]
```

Comment:
No love for device invitations ?
(I think it would be better to let the filter be done on the gui side, or to return a tuple `(<list of user invitations>, <optional device invitation>)`)

Reply:
BTW this let me to a question: wouldn't it be better to change the `invitation_list` api from this
```python
{
    "invitations": [
        <user invitation 1>,
        <device invitation>,
        <user invitation 2>,
    ]
}
```

to this:
```python
{
    "user_invitations": [<user invitation 1>, <user invitation 2>],
    "device_invitation": <device invitation>
 }
```
This would enforce the fact there could be at most only one device invitation at a time, and would avoid weird filtering on client side to sort client/device invitations

What do you think ? (and @vxgmichel )

Reply:
LGTM!